### PR TITLE
Feature/1388 list v sphere networks

### LINKED ISSUE #1388: Offer endpoint to list possible vSphere networks
### Current state
Customers who want to use a different network for each user-cluster must do this right now by using a different vm-template in vSphere.

With https://github.com/kubermatic/machine-controller/pull/240 & https://github.com/kubermatic/kubermatic/pull/1299 we now can specify a network for a new cluster.

### Desired state
To make the user select a network via the UI, the kubermatic-api need to return the list of available vSphere networks.

The implementation in the UI for this endpoint is tracked in https://github.com/kubermatic/dashboard-v2/issues/683

### PR DESCRIPTION
**Pull request by pkavajin**
**What this PR does / why we need it**:
Adds a endpoint for listing vsphere networks of a datacenter

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1388 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/api/v1/types.go
Comment (p0lyn0mial) [inline api/pkg/api/v1/types.go:241]: It seems like when working with `OpenStack` we need also an `ID` to uniquely identify the network.
See: https://github.com/kubermatic/kubermatic/blob/master/api/pkg/api/v1/types.go#L232

The `ID` is not required for `VSphere` ?

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [inline api/pkg/provider/cloud/vsphere/provider.go]: please don't return `apiv1` objects from the `VSphere` provider - it would be great if you could also change  existing methods.

see: https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/openstack.go#L106

To be honest there is no good reason for `VSphere` provider to know about `kubermaticv1.CloudSpec`. Not sure how much time do you have but it would be nice if we could change this as well.

File: api/pkg/handler/vsphere.go
Comment (p0lyn0mial) [inline api/pkg/handler/vsphere.go:15]: Please add a unit test, happy-path is good enough so that we test encoding, decoding, paths and the output.

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/handler/vsphere.go
Comment (p0lyn0mial) [inline api/pkg/handler/vsphere.go:7]: we group imports, see: https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/cluster_test.go#L3

File: api/pkg/api/v1/types.go
Comment (p0lyn0mial) [inline api/pkg/api/v1/types.go]: why did you add this type ?

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/api/v1/types.go
Comment (pkavajin) [reply api/pkg/api/v1/types.go]: Wtf, sorry, merge fail, gonna get rid of it.

File: api/pkg/api/v1/types.go
Comment (pkavajin) [reply api/pkg/api/v1/types.go:241]: vsphere uses some uuid in internal stuff, everything external works using name/path, so name should be fine as identifier for us (also, current vsphere network related voodoo uses name as well)

File: api/pkg/handler/vsphere.go
Comment (p0lyn0mial) [reply api/pkg/handler/vsphere.go:15]: it looks like @kgroschoff and @glower will add some unit test as well. Perhaps you can work together since it looks like no-one has ever written a unit test for a cloud provider.

https://github.com/kubermatic/kubermatic/pull/1516
https://github.com/kubermatic/kubermatic/pull/1512

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [inline api/pkg/provider/cloud/vsphere/provider.go]: move it before `131` and pass `ctx` to `finder.Datacenter`

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [inline api/pkg/provider/cloud/vsphere/provider.go:163]: I think that `backing` object implements the following interface

```
type BaseVirtualDeviceBackingInfo interface {
   GetVirtualDeviceBackingInfo() *VirtualDeviceBackingInfo
 }
```

it is not possible to get `DeviceName` from `VirtualDeviceBackingInfo` ?

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [inline api/pkg/provider/cloud/vsphere/provider.go]: we don't have to narrow down the search path ? 
I noticed that some calls use the `rootPath` (`getVsphereRootPath`)

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (pkavajin) [reply api/pkg/provider/cloud/vsphere/provider.go]: -> as discussed, apiv1 object is changed, kubermaticv1.* would affect all providers + interface, should be an dedicated issue

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (pkavajin) [reply api/pkg/provider/cloud/vsphere/provider.go]: thats the root path under which folders should be created for vms, networks are independent from that hierarchy

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (pkavajin) [reply api/pkg/provider/cloud/vsphere/provider.go:163]: Without explicitly casting? Dont think so, its not part of the interface.

File: api/cmd/kubermatic-api/swagger.json
Comment (p0lyn0mial) [inline api/cmd/kubermatic-api/swagger.json]: I think that you need to regenerate the swagger spec.

File: api/pkg/handler/vsphere_test.go
Comment (p0lyn0mial) [inline api/pkg/handler/vsphere_test.go:72]: was the response prepared by the `vSpehereMock` ?

is there a way to change/manipulate `vSphereMock` responses ?

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [inline api/pkg/provider/cloud/vsphere/provider.go:161]: should we log the error message ?

File: api/pkg/handler/routing_test.go
Comment (p0lyn0mial) [inline api/pkg/handler/routing_test.go:143]: why did you add this ?

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (pkavajin) [reply api/pkg/provider/cloud/vsphere/provider.go:161]: dont think so, since else we ll log errors on each api call, because its not "a real" error but a sucky way to filter out distrubuted ports and other virtual net interfaces

File: api/pkg/handler/vsphere_test.go
Comment (pkavajin) [reply api/pkg/handler/vsphere_test.go:72]: its part of the model which tearUp() initializes, .ESX creates a default model with dummy data which contains this network

File: api/cmd/kubermatic-api/swagger.json
Comment (pkavajin) [reply api/cmd/kubermatic-api/swagger.json]: aw youre absolutly right, my fail

File: api/pkg/handler/routing_test.go
Comment (pkavajin) [reply api/pkg/handler/routing_test.go:143]: Because we need a vsphere datacenter pointing to the local mock server for testing.
(To go a bit more into detail: we setup our endpoints using createTestEndpoint which gets seeded by this data, by having a genuine vsphere entry we dont have to implement any abstractions for injecting a mocked backend)

File: api/pkg/handler/vsphere_test.go
Comment (p0lyn0mial) [reply api/pkg/handler/vsphere_test.go:72]: is there a way to change this model ? for example add some data ? I'm just curious.

Comment (pkavajin) [reply api/pkg/handler/vsphere_test.go:72]: Absolutly, the simulator is the cool part of the api, giving you CRUD actions and function hooks for almost all resource types, see https://github.com/vmware/govmomi/tree/master/simulator

File: api/pkg/handler/routing_test.go
Comment (p0lyn0mial) [reply api/pkg/handler/routing_test.go:143]: sgtm.

File: api/pkg/handler/vsphere_test.go
Comment (p0lyn0mial) [reply api/pkg/handler/vsphere_test.go:72]: nice, if there is a need then we could add/modify/remove data from the simulator.

File: api/pkg/provider/cloud/vsphere/provider.go
Comment (p0lyn0mial) [reply api/pkg/provider/cloud/vsphere/provider.go:163]: yes, okay.

**Review by p0lyn0mial — APPROVED** [review]
this LGTM but I think that we need to get LGTM form other code owners.

**Review by mrIncompetent — APPROVED** [review]

File: api/pkg/handler/request.go
Comment (alvaroaleman) [inline api/pkg/handler/request.go:327]: Why does this return an interface instead of a `*http.Request`?

**Review by alvaroaleman — APPROVED** [review]

File: api/pkg/handler/request.go
Comment (alvaroaleman) [reply api/pkg/handler/request.go:327]: Nvm, found it myself

**Pull request merged**
